consensus/polybft/proposer_calculator.go

```diff
@@ -0,0 +1,499 @@
+		if err != nil {
+			return nil, err
+		}
+
+		snapshot = NewProposerSnapshot(1, genesisValidatorsSet)
+	}
+
+	return snapshot, nil
+}
+
+// NewProposerSnapshot creates ProposerSnapshot with height and validators with all priorities set to zero
+func NewProposerSnapshot(height uint64, validators []*ValidatorMetadata) *ProposerSnapshot {
+	validatorsSnap := make([]*PrioritizedValidator, len(validators))
+
+	for i, x := range validators {
+		validatorsSnap[i] = &PrioritizedValidator{Metadata: x, ProposerPriority: int64(0)}
+	}
+
+	return &ProposerSnapshot{
+		Round:      0,
+		Proposer:   nil,
+		Height:     height,
+		Validators: validatorsSnap,
+	}
+}
+
+// Gets total voting power from all the validators
+func (pcs ProposerSnapshot) GetTotalVotingPower() int64 {
+	totalVotingPower := int64(0)
+
+	for _, v := range pcs.Validators {
+		totalVotingPower = safeAddClip(totalVotingPower, int64(v.Metadata.VotingPower))
+	}
+
+	return totalVotingPower
+}
+
+// Returns copy of current ProposerSnapshot object
+func (pcs *ProposerSnapshot) Copy() *ProposerSnapshot {
+	var proposer *PrioritizedValidator
+
+	valCopy := make([]*PrioritizedValidator, len(pcs.Validators))
+
+	for i, val := range pcs.Validators {
+		valCopy[i] = &PrioritizedValidator{Metadata: val.Metadata.Copy(), ProposerPriority: val.ProposerPriority}
+
+		if pcs.Proposer != nil && pcs.Proposer.Metadata.Address == val.Metadata.Address {
+			proposer = valCopy[i]
+		}
+	}
+
+	return &ProposerSnapshot{
+		Validators: valCopy,
+		Height:     pcs.Height,
+		Round:      pcs.Round,
+		Proposer:   proposer,
+	}
+}
+
+// ProposerCalculator interface - proposer calculator algorithm should implement this interface
+type ProposerCalculator interface {
+	// CalcProposer calculates next proposer
+	CalcProposer(round, height uint64) (types.Address, error)
+
+	// Update updates ProposerSnapshot to block with number `blockNumber`
+	Update(blockNumber uint64, config *runtimeConfig, state *State) error
+
+	// GetLatestProposer returns latest calculated proposer if any
+	GetLatestProposer(round, height uint64) (types.Address, bool)
+
+	// GetSnapshot returns copy of the current snapshot
+	GetSnapshot() (*ProposerSnapshot, bool)
+}
+
+// check if interface is matching implementation
+var _ ProposerCalculator = (*proposerCalculator)(nil)
+
+type proposerCalculator struct {
+	// current snapshot
+	snapshot *ProposerSnapshot
+
+	// logger instance
+	logger hclog.Logger
+}
+
+// NewProposerCalculator creates a new proposer calculator object
+func NewProposerCalculator(config *runtimeConfig, logger hclog.Logger) (*proposerCalculator, error) {
+	snap, err := NewProposerSnapshotFromState(config, logger)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &proposerCalculator{
+		snapshot: snap,
+		logger:   logger,
+	}, nil
+}
+
+// NewProposerCalculator creates a new proposer calculator object
+func NewProposerCalculatorFromSnapshot(pcs *ProposerSnapshot, logger hclog.Logger) *proposerCalculator {
```

Comment:
So `proposerCalculator` in this scenario does two things? One is to keep the up-to-date snapshot with the calls to `Update` and another one is to serve as the "Snapshot" object that calculates the proposer.

We should not use it for both things at the same time. Move the `CalcProposer` function to the Snapshot itself and you remove the need for this two functionalities.

Reply:
Agree. Fixed https://github.com/0xPolygon/polygon-edge/pull/1018/commits/a46c6ec90b94f099b894f1e7aec34258d99b8b26